### PR TITLE
Changes on validation and lemmatisation

### PR DESCRIPTION
**Pull request by raquel-ucl**
Replaces #73 

* Validation works now for all tested files.
* Refactored to also do lemmatisation, although the ORACC server doesn't lemmatise yet (Steve is looking into it)
* Line numbers added.
* Syntax highlight for error messages returned by ORACC server added.
* Error messages shown in tooltip upon user hover.
* Other minor changes and fixes added (see diff).

### CODE REVIEW

File: python/nammu/view/ToolbarView.py [hunk 1]
```diff
@@ -42,8 +42,8 @@ def __init__(self, controller):
         tooltips['paste'] = 'Paste clipboard content'
         tooltips['validate'] = 'Check current ATF correctness'
         tooltips['lemmatise'] = 'Obtain lemmas for current ATF text'
-        tooltips['unicode'] = 'Use Unicode characters'
-        tooltips['console'] = 'View/Hide Console'
+        # tooltips['unicode'] = 'Use Unicode characters'
+        # tooltips['console'] = 'View/Hide Console'
         tooltips['displayModelView'] = 'Change to ATF data model view'
```
Comment (jenshnielsen) [inline python/nammu/view/ToolbarView.py]: Is this going back in or should it just be deleted?

File: python/nammu/SOAPClient/HTTPRequest.py [hunk 2]
```diff
@@ -14,39 +16,121 @@ def __init__(self, url, method, **kwargs):
...
+        self.rootpkg = MIMEApplication(self.envelope, 'xop+xml', encode_7or8bit)
+        self.set_multipart_payload()
+        self.document = MIMEBase('*','*')
+        self.set_document_payload(atf_basename, atf_text)
+
+        #The headers can't be created until the body is finished since they need
+        #it to populate the Content-Length header
+        self.set_multipart_headers()
+
+        print "*"*30
```
Comment (jenshnielsen) [inline python/nammu/SOAPClient/HTTPRequest.py]: I guess this should be replaced by a logging functionality but that can wait till after this PR

File: python/nammu/SOAPClient/HTTPRequest.py [hunk 3]
```diff
@@ -102,13 +186,25 @@ def get_headers(self):
...
 
     def get_body(self):
         """
         Return dict with message body/payload - ready to use by requests module.
         """
-        pass
+        headers = dict(self.mtompkg.items())
+        body = self.mtompkg.as_string().split('\n\n', 1)[1]
+        boundary = self.mtompkg.get_boundary()
+        if boundary != None:
```
Comment (jenshnielsen) [inline python/nammu/SOAPClient/HTTPRequest.py]: Add a comment explaining what this does would be useful for future use I think

File: python/nammu/SOAPClient/HTTPRequest.py [hunk 4]
```diff
@@ -14,39 +16,121 @@ def __init__(self, url, method, **kwargs):
...
+            self.mtompkg.add_header(header, value)
+
+    def set_multipart_params(self):
+        params = ['charset', 'type', 'start', 'start-info', 'boundary']
+        values = ['utf-8', 'application/xop+xml', '<SOAP-ENV:Envelope>',
+                  'application/soap+xml', '==========boundary========']
+        for param, value in zip(params, values):
+            self.mtompkg.set_param(param, value)
+    #
+    # def set_element_params(self, params, values, element):
```
Comment (jenshnielsen) [inline python/nammu/SOAPClient/HTTPRequest.py]: Should these be used or removed?

File: python/nammu/SOAPClient/SOAPClient.py [hunk 5]
```diff
@@ -33,13 +37,35 @@ def send(self):
         """
         Elaborate HTTP POST request and send it to ORACC's server.
         """
-        pass
+        url = "http://oracc.museum.upenn.edu:8085"
+        headers = dict(self.request.get_headers())
+        body = self.request.get_body()
+        self.response = requests.post(url, data=body, headers=headers)
+        # try:
```
Comment (jenshnielsen) [inline python/nammu/SOAPClient/SOAPClient.py]: Delete?

File: python/nammu/SOAPClient/SOAPClient.py [hunk 6]
```diff
@@ -33,13 +37,35 @@ def send(self):
...
+        # here: https://docs.python.org/2/library/xml.etree.elementtree.html
+        return xml_root[0][0][0][0].text
+
+    def wait_for_response(self, id):
         """
         Check for a response to the request and obtain response zip file.
         """
-        pass
+        while True:
+            ready_response = requests.get('http://oracc.museum.upenn.edu/p/' + id)
```
Comment (jenshnielsen) [inline python/nammu/SOAPClient/SOAPClient.py]: Server name should probably only be defined once? define server, port and subdir below and combine it here

File: python/nammu/SOAPClient/SOAPClient.py [hunk 7]
```diff
@@ -58,3 +84,41 @@ def create_request_zip(self):
...
+        Manipulate response to substract the content of oracc.log that is in the
+        returned binary-coded zip file.
+        """
+        self.response.content
+        binary_body = re.split('--==.*==', self.response.content)[2].split('\r\n')[5]
+
+        f = StringIO.StringIO()
+        f.write(bytearray(binary_body))
+
+        # if zipfile.is_zipfile(f):
```
Comment (jenshnielsen) [inline python/nammu/SOAPClient/SOAPClient.py]: delete

File: python/nammu/SOAPClient/SOAPClient.py [hunk 8]
```diff
@@ -58,3 +84,41 @@ def create_request_zip(self):
...
+        print "@"*30
+        print request_log
+        print "@"*30
+        if autolem:
+            print autolem
+            print "@"*30
+
+        return oracc_log, request_log, autolem
+
+    def get_autolem(self):
```
Comment (jenshnielsen) [inline python/nammu/SOAPClient/SOAPClient.py]: This function should either do something, be removed or at lease have a pass in the body

File: python/nammu/SOAPClient/HTTPRequest.py [hunk 9]
```diff
@@ -14,39 +16,121 @@ def __init__(self, url, method, **kwargs):
...
+        #Content-Transfer-Encoding is set to 7bit by default
+        del(self.rootpkg['Content-Transfer-Encoding'])
+        headers = ['Content-ID', 'Content-Transfer-Encoding']
+        values = ['<SOAP-ENV:Envelope>', 'binary']
+        for header, value in zip(headers, values):
+            self.rootpkg.add_header(header, value)
+
+    def set_multipart_headers(self):
+        headers = ['Host', 'Content-Length', 'Connection']
+        # values = [self.url, len(str(self.mtombody)), 'close']
```
Comment (jenshnielsen) [inline python/nammu/SOAPClient/HTTPRequest.py]: delete?

**Comment by jenshnielsen** [comment]
Apart from the inline comments I think this looks good to merge

File: python/nammu/test/test_soap.py [hunk 10]
```diff
@@ -4,19 +4,38 @@
...
+        client = SOAPClient('http://oracc.museum.upenn.edu:8085', method='POST')
+        client.create_request(command='atf',
+                              keys=['tests/mini', '00atf/hyphens.atf'],
+                              atf_basename='hyphens.atf',
+                              atf_text=open('resources/test/request.zip').read())
+        test_headers = client.request.get_headers()
+        assert test_headers == goal_headers
 
     @pytest.mark.xfail
-    def test_http_post_headers(self):
```
Comment (jenshnielsen) [inline python/nammu/test/test_soap.py]: Can we add a comment to explain why these are xfail?

File: python/nammu/view/ToolbarView.py (hunk 1 above)
Comment (raquel-ucl) [inline python/nammu/view/ToolbarView.py]: They'll come back, but not any time soon, so we can re-add those later on.

File: python/nammu/SOAPClient/HTTPRequest.py (hunk 2 above)
Comment (raquel-ucl) [inline python/nammu/SOAPClient/HTTPRequest.py]: Yep, there's an issue for this: #20

File: python/nammu/SOAPClient/SOAPClient.py (hunk 5 above)
Comment (raquel-ucl) [inline python/nammu/SOAPClient/SOAPClient.py]: I have to better capture the exceptions from requests. I was working on that before making the changes for this PR, so I'll remove it for now, but will be in the next PR.

File: python/nammu/SOAPClient/HTTPRequest.py (hunk 9 above)
Comment (raquel-ucl) [inline python/nammu/SOAPClient/HTTPRequest.py]: Well spotted, that should be uncommented and the line bellow deleted. Done.

File: python/nammu/test/test_soap.py (hunk 10 above)
Comment (raquel-ucl) [inline python/nammu/test/test_soap.py]: The should not xfail anymore.

**Comment by raquel-ucl** [comment]
@jenshnielsen I think I've implemented all the changes. Should be ready to merge.

**Comment by raquel-ucl** [comment]
Tests needed some update. Done now, tests pass, so ready to merge.

**Pull request merged**